ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java

```diff
@@ -245,52 +274,102 @@ public long getBytesRead() {
 
   @Override
   public byte[] toBytes() {
-    byte[] offsetBytes = offset != null ? offset.toBytes() : ZERO_LENGTH_ARRAY;
-    byte[] sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
-    byte[] incarnationIdBytes = incarnationId != null ? incarnationId.toString().getBytes() : ZERO_LENGTH_ARRAY;
-    byte[] storeKeyBytes = storeKey != null ? storeKey.toBytes() : ZERO_LENGTH_ARRAY;
-    int size = VERSION_SIZE + TYPE_SIZE;
-    if (type != Type.Uninitialized) {
-      size += INCARNATION_ID_LENGTH_SIZE + incarnationIdBytes.length + SESSION_ID_LENGTH_SIZE + sessionIdBytes.length
-          + offsetBytes.length;
-      if (type == Type.JournalBased) {
-        size += INCLUSIVE_BYTE_SIZE;
-      } else if (type == Type.IndexBased) {
-        size += storeKeyBytes.length;
-      }
-    }
-    byte[] buf = new byte[size];
-    ByteBuffer bufWrap = ByteBuffer.wrap(buf);
-    // add version
-    bufWrap.putShort(VERSION_2);
-    // add type
-    bufWrap.putShort((short) type.ordinal());
-    if (type != Type.Uninitialized) {
-      // add incarnationId
-      bufWrap.putInt(incarnationIdBytes.length);
-      bufWrap.put(incarnationIdBytes);
-      // add sessionId
-      bufWrap.putInt(sessionIdBytes.length);
-      bufWrap.put(sessionIdBytes);
-      // add offset
-      bufWrap.put(offsetBytes);
-      if (type == Type.JournalBased) {
-        bufWrap.put(getInclusive() ? (byte) 1 : (byte) 0);
-      } else if (type == Type.IndexBased) {
+    byte[] buf = null;
+    switch (version) {
+      case VERSION_0:
+        int offsetSize = 8;
+        int startOffsetSize = 8;
+        byte[] sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        byte[] storeKeyBytes = storeKey != null ? storeKey.toBytes() : ZERO_LENGTH_ARRAY;
+        int size = VERSION_SIZE + SESSION_ID_LENGTH_SIZE + sessionIdBytes.length +
+            offsetSize + startOffsetSize + storeKeyBytes.length;
+        buf = new byte[size];
+        ByteBuffer bufWrap = ByteBuffer.wrap(buf);
+        // add version
+        bufWrap.putShort(version);
+        // add sessionId
+        bufWrap.putInt(sessionIdBytes.length);
+        bufWrap.put(sessionIdBytes);
+        // add offset for journal based token
+        bufWrap.putLong((type == Type.JournalBased) ? (offset != null ? offset.getOffset() : UNINITIALIZED_OFFSET)
```

Comment:
offset cannot be null in both cases. We throw IllegalArgumentException in the constructor. Have removed the null check here.